torch/testing/_internal/common_utils.py

```diff
@@ -1302,6 +1302,30 @@ def maybeWarnsRegex(self, category, regex=''):
                         msg += '\n'
                     self.fail(msg)
 
+    @contextmanager
+    def mustWarnRegex(self, category, regex=''):
```

Comment:
I want to suggest a different name for this context manager but I'm struggling to think of one.

Let me try to breakdown how this works so I'm sure I understand it:

- if no warning is thrown it fails the test
- if the expected warning is thrown this context manager exits without issue
- if multiple are warnings then ... they ALL must be the given warning or else this fails an `assertTrue`?

Should this context manager instead verify that the warning it's looking for is thrown, but ignore all other warnings? Or should there be a mechanism to specify that multiple warnings can be thrown and that's OK?

Reply:
I will change it to `assert len(ws) == 1` instead

Reply:
OK I have a different proposal here now. What if we wrap assertWarnsRegex and always enable this setting when we do it? 

See https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertWarnsRegex

It seems like when testing we do want to validate all warnings?

Reply:
Sorry, just saw this now.

So override the same name or add a new `assertWarnsAlwaysRegex` ?

Reply:
I think just override assertWarnsRegex to call the base assertWarnsRegex but with this flag set while that happens. So like:

```
def assertWarnsRegex(...):
  try:
    # set flag
    self.super().assertWarnsRegex(...)
  finally:
    # reset flag
```


An unfortunate side-effect of this change is that the test suite will not be thread safe, but oh well.

Reply:
This doesn't work. It somehow captures the warning in a way that fails the test with "AssertionError: UserWarning not triggered". Is it critical that we re-use the super's class method? Here is what I was using"
```
    def assertWarnsRegex(self, *args, **kwargs):
        """Overrided the super class's assertWarnRegex with one that calls
        set_warn_always() to force TORCH_WARN_ONCE to behave like TORCH_WARN
        """
        prev = torch.is_warn_always_enabled()
        torch.set_warn_always(True)
        try:
            return super().assertWarnsRegex(*args, **kwargs)
        finally:
            torch.set_warn_always(prev)
```

Reply:
What you've implemented looks correct to me - would you elaborate a little on what you're seeing? Is it that assertWarnsRegex only captures the first warning if multiple are thrown, and that by enabling this flag different warnings are thrown first so these checks are failing?

If so, no we don't have to use super. What you had previously where the warnings are reviewed would be preferable. Let's not worry about the name or design tweaks if that's the case.

Reply:
Here is the complete failure:
```
    @onlyCPU
    def test_warn_always_caught(self, device):
        # Check that we can catch a TORCH_WARN_ONCE warning twice
        # since mustWarnRegex uses set_warn_always(True) which changes
        # TORCH_WARN_ONCE to TORCH_WARN
        a = np.arange(10)
        a.flags.writeable = False
        with self.assertWarnsRegex(UserWarning, '.*non-writeable.*'):
            torch.from_numpy(a)
    
        # OK, got it once, now try again
        with self.assertWarnsRegex(UserWarning, '.*non-writeable.*'):
>           torch.from_numpy(a)
E           AssertionError: UserWarning not triggered

test/test_torch.py:3291: AssertionError
```

I would prefer to not inherit the method since it only seems to complicate things.

Reply:
That's very odd. OK, let's go with the custom content manager that you had earlier for now.